packages/video_player/video_player/lib/video_player.dart

```diff
@@ -726,20 +729,46 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   ///
   /// If no [closedCaptionFile] was specified, this will always return an empty
   /// [Caption].
+
   Caption _getCaptionAt(Duration position) {
-    if (_closedCaptionFile == null) {
+    if (_closedCaptionFile == null || _sortedCaptions == null) {
       return Caption.none;
     }
 
     final Duration delayedPosition = position + value.captionOffset;
-    // TODO(johnsonmh): This would be more efficient as a binary search.
-    for (final Caption caption in _closedCaptionFile!.captions) {
-      if (caption.start <= delayedPosition && caption.end >= delayedPosition) {
-        return caption;
-      }
+
+    final int captionIndex = collection.binarySearch<Caption>(
+      _sortedCaptions!,
+      Caption(
+        number: -1,
+        start: delayedPosition,
+        end: delayedPosition,
+        text: '',
+      ),
+      compare: (Caption candidate, Caption search) {
+        if (search.start < candidate.start) {
+          return 1;
+        } else if (search.start > candidate.end) {
+          return -1;
+        } else {
+          // delayedPosition is within [candidate.start, candidate.end]
+          return 0;
+        }
+      },
+    );
+
+    // -1 means not found by the binary search.
+    if (captionIndex == -1) {
+      return Caption.none;
+    }
+
+    final Caption caption = _sortedCaptions![captionIndex];
+    // check if it really fits within that caption's [start, end].
```

Comment:
Under what circumstances will the `binarySearch` return a result that fails this check? If this is necessary, it seems like the `compare` function above is incorrect.